File: agenkit-scala/src/main/scala/io/agenkit/adapters/TokenUsage.scala
package io.agenkit.adapters

import io.agenkit.core.Message

/** Normalized, typed token usage for LLM adapter responses.
  *
  * Adapters record token counts in `Message.metadata("usage")` as a map, but
  * key names differ between the `prompt_tokens`/`completion_tokens` convention
  * and the Anthropic `input_tokens`/`output_tokens` convention.
  * [[TokenUsage.fromMessage]] normalizes both into one type so cost-metering and
  * budgeting layers consume a single shape.
  *
  * Fields are `0` when the provider does not report them. The cache fields are
  * provider-dependent (e.g. Anthropic prompt caching, including via Bedrock) and
  * are `0` when caching is inactive.
  *
  * Mirrors the Go reference (`agenkit-go/adapter/llm/usage.go`).
  */
case class TokenUsage(
  promptTokens: Long,
  completionTokens: Long,
  totalTokens: Long,
  cacheReadTokens: Long,
  cacheCreationTokens: Long
)

object TokenUsage:

  /** Extracts normalized token usage from an adapter response message.
    *
    * Reads the `metadata("usage")` map, normalizing both naming conventions
    * (`prompt_tokens`/`completion_tokens` and Anthropic's
    * `input_tokens`/`output_tokens`) and the cache keys
    * (`cache_read_tokens`/`cache_creation_tokens`, plus the raw provider aliases
    * `cache_read_input_tokens`/`cache_creation_input_tokens`).
    *
    * @return the usage, or `None` when no usage metadata is present. When
    *   `total_tokens` is absent it is derived as prompt + completion.
    */
  def fromMessage(message: Message): Option[TokenUsage] =
    message.metadata.get("usage").collect { case usage: Map[?, ?] =>
      val m = usage.asInstanceOf[Map[String, Any]]

      def pick(keys: String*): Long =
        keys.iterator.flatMap(k => m.get(k)).map(toLong).find(_ != 0L).getOrElse(0L)

      val prompt = pick("prompt_tokens", "input_tokens")
      val completion = pick("completion_tokens", "output_tokens")
      val total = pick("total_tokens") match
        case 0L => prompt + completion
        case t  => t

      TokenUsage(
        promptTokens = prompt,
        completionTokens = completion,
        totalTokens = total,
        cacheReadTokens = pick("cache_read_tokens", "cache_read_input_tokens"),
        cacheCreationTokens =
          pick("cache_creation_tokens", "cache_creation_input_tokens", "cache_write_tokens")
      )
    }

  /** Coerce a numeric metadata value to Long; 0 for non-numbers. */
  private def toLong(v: Any): Long = v match
    case n: Int    => n.toLong
    case n: Long   => n
    case n: Short  => n.toLong
    case n: Byte   => n.toLong
    case n: Double => n.toLong
    case n: Float  => n.toLong
    case _         => 0L

File: agenkit-cs/src/Agenkit/Adapters/TokenUsage.cs
using System.Text.Json;
using Agenkit.Core;

namespace Agenkit.Adapters;

/// <summary>
/// Normalized, typed token usage for LLM adapter responses.
///
/// Adapters record token counts in <c>Message.Metadata["usage"]</c>, but key
/// names differ between the <c>prompt_tokens</c>/<c>completion_tokens</c>
/// convention and the Anthropic <c>input_tokens</c>/<c>output_tokens</c>
/// convention. <see cref="FromMessage"/> normalizes both into one type so
/// cost-metering and budgeting layers consume a single shape.
///
/// Fields are 0 when the provider does not report them. The cache fields are
/// provider-dependent (e.g. Anthropic prompt caching, including via Bedrock) and
/// are 0 when caching is inactive.
///
/// Mirrors the Go reference (<c>agenkit-go/adapter/llm/usage.go</c>).
/// </summary>
public record TokenUsage(
    long PromptTokens,
    long CompletionTokens,
    long TotalTokens,
    long CacheReadTokens,
    long CacheCreationTokens)
{
    /// <summary>
    /// Extracts normalized token usage from an adapter response message.
    /// Reads the <c>Metadata["usage"]</c> entry, accepting either a nested
    /// dictionary or a <see cref="JsonElement"/> object, and normalizing both
    /// key conventions plus the cache keys (normalized and raw provider aliases).
    /// </summary>
    /// <returns>The usage, or <c>null</c> when no usage metadata is present.
    /// When <c>total_tokens</c> is absent it is derived as prompt + completion.</returns>
    public static TokenUsage? FromMessage(Message? message)
    {
        if (message?.Metadata is null || !message.Metadata.TryGetValue("usage", out var raw))
            return null;

        var lookup = MakeLookup(raw);
        if (lookup is null)
            return null;

        long Pick(params string[] keys)
        {
            foreach (var key in keys)
            {
                var v = lookup(key);
                if (v.HasValue)
                    return v.Value;
            }
            return 0;
        }

        var prompt = Pick("prompt_tokens", "input_tokens");
        var completion = Pick("completion_tokens", "output_tokens");
        var total = Pick("total_tokens");
        if (total == 0)
            total = prompt + completion;

        return new TokenUsage(
            prompt,
            completion,
            total,
            Pick("cache_read_tokens", "cache_read_input_tokens"),
            Pick("cache_creation_tokens", "cache_creation_input_tokens", "cache_write_tokens"));
    }

    /// <summary>
    /// Builds a key→long lookup over the usage value, handling both a plain
    /// dictionary (counts stored directly) and a JsonElement object (raw parse).
    /// Returns null when the value is neither.
    /// </summary>
    private static Func<string, long?>? MakeLookup(object raw)
    {
        switch (raw)
        {
            case JsonElement json when json.ValueKind == JsonValueKind.Object:
                return key =>
                    json.TryGetProperty(key, out var el) && el.ValueKind == JsonValueKind.Number
                        ? el.GetInt64()
                        : null;
            case IReadOnlyDictionary<string, object> dict:
                return key => dict.TryGetValue(key, out var v) ? ToLong(v) : null;
            default:
                return null;
        }
    }

    /// <summary>Coerce a numeric value to long; null for non-numbers.</summary>
    private static long? ToLong(object? v) => v switch
    {
        long l => l,
        int i => i,
        short s => s,
        byte b => b,
        double d => (long)d,
        float f => (long)f,
        JsonElement je when je.ValueKind == JsonValueKind.Number => je.GetInt64(),
        _ => null
    };
}
